### DISCUSSION TITLE
Migration to conform and nvim-lint

### DISCUSSION THREAD

**Discussion by Jomik** [LazyVim/LazyVim#1522] [Q&A] [answered]
I see @folke has chosen to migrate formatting to [conform.nvim](https://github.com/stevearc/conform.nvim) and linting to [nvim-lint](https://github.com/mfussenegger/nvim-lint).
Nice!

I am unsure how I would migrate to use them though, as I still see a lot of `none-ls` even if I enable `conform` and `nvim-lint`. Won't that result in two formatters and linters running?

I also got issues with `nvim-lint` when opening markdown files, seems to be that it expects `vale` to be available, but we do not install that with `mason` anywhere.
I think we need to use mason to install the linters more aggresively, like for `docker`
https://github.com/LazyVim/LazyVim/blob/886d49eb89cc96dd8942182d4c3d87b252500831/lua/lazyvim/plugins/extras/lang/docker.lua#L28-L36
I expect this works in testing, because most people have `hadolint` installed from the `null-ls` time :)

Is there any way we can help setting things up for these two new plugins, so that we can get rid of `none-ls` fully, assuming that is the purpose?

**Comment by folke** [comment, marked as answer]
Best to disable none-ls if you want to use conform and nvim-lint instead:

```lua
  {
    "nvimtools/none-ls.nvim",
    enabled = false,
}
```

Where do you get an error about `vale` for markdown? I don't have that issue.

**Reply by folke** [reply]
I see what you mean with hadloint. Will update the spec

**Reply by Jomik** [reply]
I am just in a markdown file, trying to save. I do not have `vale` installed.
![image](https://github.com/LazyVim/LazyVim/assets/699655/1326a29c-5e0a-4f5d-8940-173f809404c2)

**Reply by Jomik** [reply]
Well, also getting the same error for `jsonlint` when opening a json file 🤔  Looks like it is either trying to run something that it shouldn't, or we are missing some dependencies in mason.
I am not mentioning `vale` or `jsonlint` anywhere in my spec.
There seems to be some defaults declared though https://github.com/mfussenegger/nvim-lint/blob/master/lua/lint.lua#L42-L53

**Reply by folke** [reply]
But LazyVim does not configre `nvim-lint` to use `vale` for markdown.

Are you sure you didn't add that somewhere in your personal config?

It's the example config on the `nvim-lint` readme. But LazyVim does not use that at all.

**Reply by folke** [reply]
Nevermind, just saw that nvim-lint configures default linters. Will remove those.

**Reply by Jomik** [reply]
Yeah - going to add an issue for `nvim-lint` on that. Hidden defaults 😄 
Thank you!

**Reply by LarryMarzanJr** [reply]
@folke to disable none-ls where do I add this? 
```lua
{
    "nvimtools/none-ls.nvim",
    enabled = false,
}
```

**Reply by fredrikaverpil** [reply]
@LarryMarzanJr just add e.g. a disabled.lua file in your `plugins` folder. You can put it in any lua file that gets read as part of [this](https://github.com/LazyVim/starter/blob/main/lua/config/lazy.lua#L18) directive.

But I think none-ls is disabled by default... so you might not have to do that at all, unless you have none-ls config somewhere in your `plugins` dir.

**Reply by dpetka2001** [reply]
My guess is that he has some previous spec configuration about `none-ls` and is getting the error that he should either remove the spec from his configuration or enable the Extra. He should search in his configuration about anything related to `none-ls` and just remove it.

**Comment by folke** [comment]
To better explain, all `none-ls` references throughout the code uses `optional = true`. 
This means that `lazy.nvim` will only add those spec fragments to the final spec if the plugin has been defined somewhere without `optional = true`.

**Reply by Jomik** [reply]
Now? https://github.com/folke/lazy.nvim/commit/f3c7169dd65f5ae528b6c930492359971014290b :D Sweet

**Reply by folke** [reply]
`optional` has been around for a couple of months now, so that already worked for this use-case.
The update I did today to `lazy.nvim` just makes it behave more in line with what users would expect.

**Reply by Jomik** [reply]
Yah, I remember. I was confused why I still saw `none-ls`, but I found a rogue reference in my spec.

**Reply by dpetka2001** [reply]
> To better explain, all none-ls references throughout the code uses optional = true

@folke Should someone also adopt `optional` in his personal spec? I mean, I also have different spec fragments of former `null-ls` references (which with the `renames` shouldn't be a problem I believe) in my configuration. I have a central `plugins/lsp.lua` and the rest fragmented across `extras/lang` directory for different languages (following the file structure hierarchy of Lazyvim). And if someone completely disables `none-ls` with `enabled = false`, I guess this doesn't matter either way?

**Reply by folke** [reply]
@dpetka2001 no need to use `optional = true` in your own specs. That's mostly useful for Neovim distros like `LazyVim`.
`AstroNvim` uses this a lot as well.

> And if someone completely disables none-ls with enabled = false, I guess this doesn't matter either way?

Exactly :)

**Comment by OneOfOne** [comment]
We still need something for code actions, I'm not sure why we'd switch off none-ls if it's maintained.

**Reply by folke** [reply]
I don't use none-ls code-actions myself. There's also currently only one place in LazyVim that uses specific none-ls code-actions. (that will stay as is, since it's an optional dependency anyway)

The plan right now is:
- have both `nvim-lint` and `conform.nvim` as extras
- keep `none-ls` as the default

After a couple of weeks, I'm probably going to switch that around, so `nvim-lint/conform.nvim` as the default and `none-ls` as an extra.

**Reply by fredrikaverpil** [reply]
> After a couple of weeks, I'm probably going to switch that around, so `nvim-lint/conform.nvim` as the default and `none-ls` as an extra.

I like being able to quickly disable formatting or diagnostics in the `<leader>u` menu. Does this work today with nvim-lint and conform.nvim or will that work only once you make none-ls into an extras?

**Reply by folke** [reply]
@fredrikaverpil yes

**Comment by fredrikaverpil** [comment]
I figured I'd share what I have so far, in terms of migrating from nulll-ls/none-ls to conform.nvim and nvim-lint: https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/lsp.lua

Works great so far! 😄 

I noticed that when printing `opts.formatters_by_ft` after extending it with my config, it looks kind of strange. Note the inconsistencies in double vs single angel brackets. Wherever I specified formatters, single angel brackets are used. But formatting all works as expected. Not sure what's up with that:

```
21:32:55 msg_show.echo formatters_by_ft: {
  css = { { "prettierd", "prettier" } },
  fish = { "fish_indent" },
  go = { "gofumpt", "goimports" },
  graphql = { { "prettierd", "prettier" } },
  handlebars = { { "prettierd", "prettier" } },
  html = { { "prettierd", "prettier" } },
  javascript = { { "prettierd", "prettier" } },
  javascriptreact = { { "prettierd", "prettier" } },
  json = { { "prettierd", "prettier" } },
  jsonc = { { "prettierd", "prettier" } },
  less = { { "prettierd", "prettier" } },
  lua = { "stylua", "stylua" },
  markdown = { { "prettierd", "prettier" } },
  ["markdown.mdx"] = { { "prettierd", "prettier" } },
  protobuf = { "buf" },
  python = { "isort", "black" },
  rust = { "rustfmt" },
  scss = { { "prettierd", "prettier" } },
  sh = { "shfmt", "shfmt" },
  terraform = { "terraform_fmt" },
  ["terraform-vars"] = { "terraform_fmt" },
  tf = { "terraform_fmt" },
  typescript = { { "prettierd", "prettier" } },
  typescriptreact = { { "prettierd", "prettier" } },
  vue = { { "prettierd", "prettier" } },
  yaml = { { "prettierd", "prettier" } }
}
```

**Reply by folke** [reply]
That's a feature and it's on purpose. Check the conform docs to know what that is. Single brackets will run all formatters. Double only the first that is available on the system
